File: 744/find-smallest-letter-greater-than-target.cpp
class Solution {
public:
    char nextGreatestLetter(vector<char>& letters, char target) {

        //陣列 letters[i]
        //for迴圈 for(int i=0; i<lettersSize; i++)
        //if判斷 if(letters[i]>target)
        //你完成了 90%, 但是有 一半的人會錯(有陷阱)

        //哨兵的技巧 sentinel 保護好你的程式,簡單,漂亮,不會錯
        char ans = 'z'+1; //字母國裡很大的數字,不合理的數字
        for(int i=0; i<letters.size(); i++){
            if(letters[i]>target && letters[i]<ans ){
            //要比 target大, 且  要比 ans小 (你要是最小的)
                ans = letters[i];
            }
        }

        if(ans=='z'+1) return letters[0];
        else return ans;        
    }
};

File: 744/find-smallest-letter-greater-than-target.py
# LeetCode 744. Find Smallest Letter Greater Than Target
# 在 letters 字母的陣列裡，找比 target 字母大的最小字母
class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        ans = letters[0]  # 如果找不到，就回傳 letters 的第1個字母
        letters.sort()  # 排序，以便逐一找答案
        for c in letters:  # 小到大，依序比較
            if c > target: return c  # 找到第1個比 target 大的字母，成功
        # 但如果都沒找到的話
        return ans  # 把一開始記起來的 letters[0] 拿來用
